File: NRDB/DrawSimulatorViewController.m
//
//  DrawSimulatorViewController.m
//  NRDB
//
//  Created by Gereon Steffens on 15.02.14.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import "DrawSimulatorViewController.h"
#import "Deck.h"
#import "CardImageViewPopover.h"
#import "Hypergeometric.h"
#import "CardThumbView.h"

@interface DrawSimulatorViewController ()
@property Deck* deck;
@property NSMutableArray* cards;    // cards in deck
@property NSMutableArray* draw;     // cards drawn
@property NSMutableArray* played;   // card's played state (BOOL)
@end

static NSInteger viewMode;

@implementation DrawSimulatorViewController

+(void) showForDeck:(Deck*)deck inViewController:(UIViewController*)vc
{
    DrawSimulatorViewController* davc = [[DrawSimulatorViewController alloc] initWithDeck:deck];
    
    [vc presentViewController:davc animated:NO completion:nil];
}

- (id)initWithDeck:(Deck*)deck
{
    self = [super initWithNibName:@"DrawSimulatorViewController" bundle:nil];
    if (self)
    {
        self.deck = deck;
        self.modalPresentationStyle = UIModalPresentationFormSheet;
    }
    return self;
}

-(void) dealloc
{
    self.tableView.delegate = nil;
    self.tableView.dataSource = nil;
    self.tableView = nil;
    
    self.collectionView.delegate = nil;
    self.collectionView.dataSource = nil;
    self.collectionView = nil;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self initCards:YES];
    
    self.titleLabel.text = l10n(@"Draw Simulator");
    [self.clearButton setTitle:l10n(@"Clear") forState:UIControlStateNormal];
    [self.doneButton setTitle:l10n(@"Done") forState:UIControlStateNormal];
    
    [self.selector setTitle:l10n(@"All") forSegmentAtIndex:6];
    self.selector.apportionsSegmentWidthsByContent = YES;
    
    self.viewModeControl.selectedSegmentIndex = viewMode;
    
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    self.tableView.hidden = viewMode == 0;
    
    [self.collectionView registerNib:[UINib nibWithNibName:@"CardThumbView" bundle:nil] forCellWithReuseIdentifier:@"cardThumb"];
    self.collectionView.hidden = viewMode == 1;
    
    // each view needs its own long press recognizer
    [self.tableView addGestureRecognizer:[[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(longPress:)]];
    [self.collectionView addGestureRecognizer:[[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(longPress:)]];
    
    self.collectionView.dataSource = self;
    self.collectionView.delegate = self;
}

-(void) initCards:(BOOL)drawInitialHand
{
    self.draw = [NSMutableArray array];
    self.cards = [NSMutableArray array];
    self.played = [NSMutableArray array];
    
    for (CardCounter* cc in self.deck.cards)
    {
        for (int c=0; c<cc.count; ++c)
        {
            [self.cards addObject:cc.card];
        }
    }
    NSAssert(self.cards.count == self.deck.size, @"size mismatch");
    
    // shuffle the cards using Fisher-Yates
    u_int32_t count = (u_int32_t)self.cards.count;
    for (int i = 0; i < count; ++i)
    {
        // Select a random element between i and end of array to swap with.
        u_int32_t n = arc4random_uniform(count - i) + i;
        [self.cards exchangeObjectAtIndex:i withObjectAtIndex:n];
    }
    
    self.oddsLabel.text = @"";
    self.drawnLabel.text = @"";
    if (drawInitialHand)
    {
        int handSize = 5;
        if ([self.deck.identity.code isEqualToString:ANDROMEDA])
        {
            handSize = 9;
        }
        [self drawCards:handSize];
    }
}

#pragma mark buttons

-(void) done:(id)sender
{
    [self dismissViewControllerAnimated:NO completion:nil];
}

-(void) clear:(id)sender
{
    [self initCards:NO];
    [self.tableView reloadData];
    [self.collectionView reloadData];
}

-(void) viewModeChange:(UISegmentedControl*)sender
{
    viewMode = sender.selectedSegmentIndex;
    self.tableView.hidden = viewMode == 0;
    self.collectionView.hidden = viewMode == 1;
}

-(void) draw:(UISegmentedControl*)sender
{
    // segments are: 0=1, 1=2, 2=3, 3=4, 4=5, 5=9, 6=All
    NSInteger numCards;
    switch (sender.selectedSegmentIndex)
    {
        case 5:
            numCards = 9;
            break;
        case 6:
            numCards = self.deck.size;
            break;
        default:
            numCards = sender.selectedSegmentIndex + 1;
            break;
    }
    
    [self drawCards:numCards];
}

-(void) drawCards:(NSInteger)numCards
{
    for (int i=0; i<numCards; ++i)
    {
        if (self.cards.count > 0)
        {
            Card* card = [self.cards objectAtIndex:0];
            [self.cards removeObjectAtIndex:0];
            [self.draw addObject:card];
            [self.played addObject:@(NO)];
        }
    }
    
    NSAssert(self.draw.count == self.played.count, @"size mismatch");
    
    NSUInteger drawn = self.draw.count;
    self.drawnLabel.text = [NSString stringWithFormat:l10n(@"%ld %@ drawn"), (unsigned long)drawn, drawn == 1 ? l10n(@"Card") : l10n(@"Cards") ];
    
    [self.tableView reloadData];
    [self.collectionView reloadData];
    
    // scroll down if not all cards were drawn
    if (numCards != self.deck.size && self.draw.count > 0)
    {
        NSIndexPath* indexPath = [NSIndexPath indexPathForRow:self.draw.count-1 inSection:0];
        [self.tableView scrollToRowAtIndexPath:indexPath atScrollPosition:UITableViewScrollPositionBottom animated:NO];
        [self.collectionView scrollToItemAtIndexPath:indexPath atScrollPosition:UICollectionViewScrollPositionBottom animated:NO];
    }
    
    // calculate drawing odds
    NSString* odds = [NSString localizedStringWithFormat:l10n(@"Odds to draw a card: 1×%.1f%%  2×%.1f%%  3×%.1f%%"),
                      [self oddsFor:1], [self oddsFor:2], [self oddsFor:3] ];
    self.oddsLabel.text = odds;
}

-(double) oddsFor:(int)cardsInDeck
{
    return 100.0 * [Hypergeometric getProbabilityFor:1 cardsInDeck:self.deck.size desiredCardsInDeck:cardsInDeck cardsDrawn:self.draw.count];
}

#pragma mark tableview

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.draw.count;
}

-(UITableViewCell*) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString* cellIdentifier = @"drawCell";
    
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    
    Card* card = [self.draw objectAtIndex:indexPath.row];
    cell.textLabel.text = card.name;
    cell.textLabel.textColor = [self.played[indexPath.row] boolValue] ? [UIColor lightGrayColor] : [UIColor blackColor];
    
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    Card* card = self.draw[indexPath.row];
    
    CGRect rect = [self.tableView rectForRowAtIndexPath:indexPath];
    [CardImageViewPopover showForCard:card fromRect:rect inView:self.tableView];
}

#pragma mark collectionview

-(UICollectionViewCell*) collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString* cellIdentifier = @"cardThumb";
    Card *card = [self.draw objectAtIndex:indexPath.row];
    
    CardThumbView* cell = [collectionView dequeueReusableCellWithReuseIdentifier:cellIdentifier forIndexPath:indexPath];
    
    cell.card = card;
    cell.imageView.layer.opacity = [self.played[indexPath.row] boolValue] ? .5 : 1;
    
    return cell;
}

-(void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    Card* card = self.draw[indexPath.row];
    UICollectionViewCell* cell = [collectionView cellForItemAtIndexPath:indexPath];
    
    // convert to on-screen coordinates
    CGRect rect = [collectionView convertRect:cell.frame toView:self.collectionView];
    
    [CardImageViewPopover showForCard:card fromRect:rect inView:self.collectionView];
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    return CGSizeMake(160, 119);
}

- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout insetForSectionAtIndex:(NSInteger)section
{
    return UIEdgeInsetsZero;
}

-(NSInteger) numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return 1;
}

-(NSInteger) collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return self.draw.count;
}

#pragma mark longpress

-(void) longPress:(UIGestureRecognizer*)gesture
{
    if (gesture.state == UIGestureRecognizerStateBegan)
    {
        NSIndexPath* indexPath;
        if (self.tableView.hidden)
        {
            CGPoint point = [gesture locationInView:self.collectionView];
            indexPath = [self.collectionView indexPathForItemAtPoint:point];
        }
        else
        {
            CGPoint point = [gesture locationInView:self.tableView];
            indexPath = [self.tableView indexPathForRowAtPoint:point];
        }
        
        if (indexPath)
        {
            BOOL played = [[self.played objectAtIndex:indexPath.row] boolValue];
            self.played[indexPath.row] = @(!played);
            NSArray* paths = @[ indexPath ];
            [self.tableView reloadRowsAtIndexPaths:paths withRowAnimation:UITableViewRowAnimationNone];
            [self.collectionView reloadItemsAtIndexPaths:paths];
        }
    }
}

@end

File: NetDeck/DrawSimulatorViewController.swift
//
//  DrawSimulatorViewController.swift
//  NetDeck
//
//  Created by Gereon Steffens on 13.12.15.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

import UIKit

class DrawSimulatorViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
    @IBOutlet weak var viewModeControl: UISegmentedControl!
    @IBOutlet weak var titleLabel: UILabel!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var collectionView: UICollectionView!
    @IBOutlet weak var drawnLabel: UILabel!
    @IBOutlet weak var oddsLabel: UILabel!
    
    @IBOutlet weak var clearButton: UIButton!
    @IBOutlet weak var doneButton: UIButton!
    @IBOutlet weak var selector: UISegmentedControl!
    
    var deck: Deck!
    var cards = [Card]()
    var drawn = [Card]()
    var played = [Bool]()
    
    static var viewMode = 0
    
    class func showForDeck(_ deck: Deck, inViewController vc: UIViewController) {
        let dvw = DrawSimulatorViewController()
        dvw.deck = deck
        dvw.modalPresentationStyle = .formSheet
        vc.present(dvw, animated: false, completion: nil)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        Analytics.logEvent("Draw Sim", attributes: ["Device": "iPad"])
        
        self.initCards(true)
        
        self.titleLabel.text = "Draw Simulator".localized()
        self.clearButton.setTitle("Clear".localized(), for: UIControlState())
        self.doneButton.setTitle("Done".localized(), for: UIControlState())
        
        self.selector.setTitle("All".localized(), forSegmentAt:6)
        self.selector.apportionsSegmentWidthsByContent = true
        
        self.viewModeControl.selectedSegmentIndex = DrawSimulatorViewController.viewMode
        
        self.tableView.tableFooterView = UIView(frame:CGRect.zero)
        self.tableView.isHidden = DrawSimulatorViewController.viewMode == 0
        
        self.collectionView.register(UINib(nibName:"CardThumbView", bundle:nil),forCellWithReuseIdentifier:"cardThumb")
        self.collectionView.isHidden = DrawSimulatorViewController.viewMode == 1
        
        // each view needs its own long press recognizer
        self.tableView.addGestureRecognizer(UILongPressGestureRecognizer(target:self, action:#selector(DrawSimulatorViewController.longPress(_:))))
        self.collectionView.addGestureRecognizer(UILongPressGestureRecognizer(target:self, action:#selector(DrawSimulatorViewController.longPress(_:))))
        
        self.collectionView.dataSource = self
        self.collectionView.delegate = self
        
        self.oddsLabel.font = UIFont.monospacedDigitSystemFont(ofSize: 17, weight: UIFontWeightRegular)
        self.drawnLabel.font = UIFont.monospacedDigitSystemFont(ofSize: 15, weight: UIFontWeightRegular)
    }
    
    func initCards(_ drawInitialHand: Bool) {
        self.drawn = [Card]()
        self.cards = [Card]()
        self.played = [Bool]()
        
        for cc in self.deck.cards {
            for _ in 0 ..< cc.count {
                cards.append(cc.card)
            }
        }
        
        assert(self.cards.count == self.deck.size, "oops")
        
        self.cards.shuffle()
        
        self.oddsLabel.text = ""
        self.drawnLabel.text = ""
        
        if drawInitialHand {
            var handSize = 5
            if self.deck.identity?.code == Card.andromeda {
                handSize = 9
            }
            self.drawCards(handSize)
        }
    }
    
    func drawCards(_ numCards: Int) {
        for _ in 0..<numCards {
            if self.cards.count > 0 {
                let card = self.cards[0]
                cards.removeFirst()
                self.drawn.append(card)
                self.played.append(false)
            }
        }
        
        assert(self.drawn.count == self.played.count, "oops")
        
        let drawn = self.drawn.count
        self.drawnLabel.text = String(format:"%ld %@ drawn".localized(),
            drawn, drawn==1 ? "Card".localized() : "Cards".localized())
        
        self.tableView.reloadData()
        self.collectionView.reloadData()
        
        // scroll down if not all cards were drawn
        if numCards != self.deck.size && self.drawn.count > 0 {
            let indexPath = IndexPath(row: self.drawn.count-1, section: 0)
            self.tableView.scrollToRow(at: indexPath, at: .bottom, animated: false)
            self.collectionView.scrollToItem(at: indexPath, at: .bottom, animated: false)
        }
        
        // calculate drawing odds
        let fmt = "Odds to draw a card: 1×%.2f%%  2×%.2f%%  3×%.2f%%".localized()
        let odds = String(format: fmt, self.oddsFor(1), self.oddsFor(2), self.oddsFor(3))
        self.oddsLabel.text = odds
    }
    
    func oddsFor(_ cards: Int) -> Double {
        let odds = 100.0 * Hypergeometric.getProbabilityFor(1, cardsInDeck:self.deck.size, desiredCardsInDeck:min(self.deck.size, cards), cardsDrawn:self.drawn.count)
        return odds
    }
    
    // MARK: table view
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.drawn.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "drawCell"
        var cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: cellIdentifier)
            cell!.selectionStyle = .none
        }
        
        let card = self.drawn[indexPath.row]
        cell!.textLabel?.text = card.name
        cell!.textLabel?.textColor = self.played[indexPath.row] ? UIColor.lightGray : UIColor.black
        
        return cell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let card = self.drawn[indexPath.row]
        
        let rect = self.tableView.rectForRow(at: indexPath)
        
        CardImageViewPopover.show(for: card, from: rect, in:self, subView:self.tableView)
    }
    
    // MARK: collection view
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.drawn.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cardThumb", for: indexPath) as! CardThumbView
        
        let card = self.drawn[indexPath.row]
        cell.card = card
        cell.imageView.layer.opacity = self.played[indexPath.row] ? 0.5 : 1
        
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: 160, height: 119)
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let card = self.drawn[indexPath.row]
        let cell = collectionView.cellForItem(at: indexPath)
        let rect = collectionView.convert(cell!.frame, to: collectionView)
        
        CardImageViewPopover.show(for: card, from: rect, in:self, subView:self.collectionView)
    }
    
    // MARK: long press
    
    func longPress(_ gesture: UIGestureRecognizer) {
        if gesture.state == .began {
            var indexPath: IndexPath?
            if self.tableView.isHidden {
                let point = gesture.location(in: self.collectionView)
                indexPath = self.collectionView.indexPathForItem(at: point)
            } else {
                let point = gesture.location(in: self.tableView)
                indexPath = self.tableView.indexPathForRow(at: point)
            }
            
            if let indexPath = indexPath {
                let played = self.played[indexPath.row]
                self.played[indexPath.row] = !played
                let paths = [ indexPath ]
                
                self.tableView.reloadRows(at: paths, with: .none)
                self.collectionView.reloadItems(at: paths)
            }
        }
    }
    
    // MARK: button actions
    
    @IBAction func done(_ sender: UIButton) {
        self.dismiss(animated: false, completion: nil)
    }
    
    @IBAction func clear(_ sender: UIButton) {
        self.initCards(false)
        self.tableView.reloadData()
        self.collectionView.reloadData()
    }
    
    @IBAction func draw(_ sender: UISegmentedControl) {
        // segments are: 0=1, 1=2, 2=3, 3=4, 4=5, 5=9, 6=All
        var numCards = 0
        switch (sender.selectedSegmentIndex) {
        case 5:
            numCards = 9
        case 6: // all
            numCards = self.deck.size
        default:
            numCards = sender.selectedSegmentIndex + 1
        }
        
        self.drawCards(numCards)
    }
    
    @IBAction func viewModeChange(_ sender: UISegmentedControl) {
        DrawSimulatorViewController.viewMode = sender.selectedSegmentIndex
        self.tableView.isHidden = DrawSimulatorViewController.viewMode == 0
        self.collectionView.isHidden = DrawSimulatorViewController.viewMode == 1
    }
}
